### PR TITLE
move the 4th order driver to C++

### PR DESCRIPTION
**Pull request by zingale**
## PR checklist

- [ ] test suite needs to be run on this PR
- [ ] this PR will change answers in the test suite
- [ ] all functions have docstrings as per the coding conventions
- [ ] the `CHANGES` file has been updated
- [ ] if appropriate, this change is described in the docs

### CODE REVIEW

**Comment by zingale** [comment]
This seems to work on CPUs.  All tests are passing.  I haven't yet marked it up for GPUs, but I put in the elixirs (although, I am not completely certain I am doing them correctly inside the loop over direction).  Once this new C++ driver is merged, I'll work on the GPUing.  That will then finish issue #526

**Comment by zingale** [comment]
tests passed:
http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/2020-01-21-002/index.html

File: Source/hydro/Castro_hydro_F.H [hunk 1]
```diff
@@ -434,8 +434,8 @@ extern "C"
 
   void ca_mol_consup
```
Comment (maxpkatz) [inline Source/hydro/Castro_hydro_F.H]: I'm confused by this interface change to ca_mol_consup; it doesn't look like you're passing shk in the C++ call to this function.

Comment (zingale) [reply Source/hydro/Castro_hydro_F.H]: I'm confused too.... I'll fix this

**Comment by zingale** [comment]
all SDC tests pass
http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/2020-01-25/index.html

**Pull request merged**